### PR TITLE
update test credentials validation

### PR DESCRIPTION
**Pull request by miha1**

### CODE REVIEW

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 1]
```diff
@@ -59,39 +61,93 @@ class WirecardAjaxController extends ModuleAdminController
...
+    }
 
-                if (('creditcard' === $method) && UrlConfigurationChecker::isUrlConfigurationValid($baseUrl, $wppUrl)) {
-                    $message = $this->l('warning_credit_card_url_mismatch');
-                }
+    /**
+     *
+     * Validate base Url.
+     * It shouldn't have any path on the end of Url
+     * @param $baseUrl
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: add the param type string

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 2]
```diff
@@ -59,39 +61,93 @@ class WirecardAjaxController extends ModuleAdminController
...
+     * @since 2.1.0
+     */
+    protected function validateCredentials($transactionService)
+    {
+        try {
+            $success = $transactionService->checkCredentials();
+        } catch (\Http\Client\Exception $exception) {
+            $success = false;
+        }
+        $status = $success ? 'ok' : 'error';
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: Why is this line needed you just send true or false forward, i think its just mapping a true false to something else that's same.

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 3]
```diff
@@ -59,39 +61,93 @@ class WirecardAjaxController extends ModuleAdminController
...
+
+    /**
+     * Send response
+     * @since 2.1.0
+     */
+    protected function sendResponse($status, $message)
+    {
+        $content = json_encode(['status' => htmlspecialchars($status), 'message' => htmlspecialchars($message)]);
+        $response = new Response($content);
+        $response->sendContent();
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: is this correct now? i think you said to me there is a function that sends the response without header is it this?

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 4]
```diff
@@ -59,39 +61,93 @@ class WirecardAjaxController extends ModuleAdminController
...
+        if (('creditcard' === $method) && !UrlConfigurationChecker::isUrlConfigurationValid($baseUrl, $wppUrl)) {
+            $message = $this->l('warning_credit_card_url_mismatch');
+            $status = 'error';
+            $this->sendResponse($status, $message);
         }
     }
+
+    /**
+     * Send response
+     * @since 2.1.0
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: params are missing

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 5]
```diff
@@ -59,39 +61,93 @@ class WirecardAjaxController extends ModuleAdminController
...
 
-                die(\Tools::jsonEncode(
-                    [
-                        'status' => htmlspecialchars($status),
-                        'message' => htmlspecialchars($message)
-                    ]
-                ));
+    /**
+     * Check if base url and wpp url are on the same level (test or production)
+     * @since 2.1.0
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: params and return are missing

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php (hunk 3 above)
Comment (miha1) [reply wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: Yes, that is the function

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php (hunk 2 above)
Comment (miha1) [reply wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: Its for status information on admin page

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 6]
```diff
@@ -59,39 +61,104 @@ class WirecardAjaxController extends ModuleAdminController
...
+
+    /**
+     * Check if base url and wpp url are on the same level (test or production)
+     * @param string $method
+     * @param string $baseUrl
+     * @param string $wppUrl
+     * @throws InvalidArgumentException
+     * @since 2.1.0
+     */
+    protected function validateUrlConfiguration($method, $baseUrl, $wppUrl)
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: This method is not returning anything why is then the massage set?

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 7]
```diff
@@ -59,39 +61,104 @@ class WirecardAjaxController extends ModuleAdminController
...
-                
-                $config = new Config($baseUrl, $httpUser, $httpPass);
-                $transactionService = new TransactionService($config, new Logger());
+    /**
+     * Test filled in credentials.
+     * @since 2.1.0
+     */
+    protected function testCredentials()
+    {
+        $method = Tools::getValue('method');
```
Comment (tomazpu) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: put this into a separate method called getPaymentMethodCode or so and in there you can do the override of sofort

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php (hunk 6 above)
Comment (miha1) [reply wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: If the url is not valid the translated message will be send back in testCredentials function
` catch (InvalidArgumentException $exception) {
            $this->sendResponse('error', $exception->getMessage());
        }`

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 8]
```diff
@@ -60,37 +65,106 @@ class WirecardAjaxController extends ModuleAdminController
...
+     * @param string $method
+     *
+     * @return bool
+     * @throws \Http\Client\Exception
+     * @since 2.1.0
+     */
+    private function validatePaymentMethod($base_url, $http_user, $http_pass, $wpp_url, $method)
+    {
+        $status = $this->validateBaseUrl($base_url) && $this->validateCredentials($base_url, $http_user, $http_pass);
+        if ($method == 'creditcard' && $status) {
```
Comment (pretzelhands) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: Here we can use `PaymentCreditCard::TYPE` to avoid using magic strings for payment types.

File: wirecardpaymentgateway/controllers/admin/WirecardAjax.php [hunk 9]
```diff
@@ -60,37 +65,106 @@ class WirecardAjaxController extends ModuleAdminController
...
+     * @param string $http_pass
+     *
+     * @return boolean
+     * @throws \Http\Client\Exception
+     * @since 2.1.0
+     */
+    private function validateCredentials($base_url, $http_user, $http_pass)
+    {
+        $config = new Config($base_url, $http_user, $http_pass);
+        $transactionService = new TransactionService($config, new Logger());
```
Comment (pretzelhands) [inline wirecardpaymentgateway/controllers/admin/WirecardAjax.php]: This is the only variable in this file still using `camelCase`. Please change it to `snake_case` to be in sync with the rest of the file :)
